### PR TITLE
[Intel XPU] Fix DP-attention `reduce_scatterv` / `all_gatherv` on XPU (XCCL)

### PR DESCRIPTION
**Pull request by Amrutha-M05**
## Motivation

The `reduce_scatterv` and `all_gatherv` collectives in `GroupCoordinator` are hard-wired to `pynccl_comm` and assume an NCCL backend. On Intel XPU (XCCL), that path isn't usable — `pynccl_comm` is disabled, so any DP-attention flow that reaches these collectives fails. This PR adds XPU-native fallbacks so DP-attention works end-to-end on XPU while keeping the existing CUDA/NCCL path untouched.

## Modifications

Single-file change in `python/sglang/srt/distributed/parallel_state.py`. Both additions are gated behind `if _is_xpu:` and taken *before* the existing `pynccl_comm.change_state(...)` block, so CUDA/ROCm/NPU behavior is byte-identical to `main`.

- **`reduce_scatterv` (XPU path):** implement `reduce_scatter` as `all_reduce` followed by a per-rank slice into `output`. Handles both the uniform-chunk case and the variable `sizes` case (offset computed as `sum(sizes[:rank])`). Allocates the output buffer if the caller didn't pass one, matching the NCCL path's contract.
- **`all_gatherv` (XPU path):** fuse `N` per-rank broadcasts inside `torch.distributed._coalescing_manager` so XCCL sees a single coalesced op instead of `N` separate broadcasts. Each rank pre-copies its own shard into its slot in the output buffer (so every rank passes the same buffer into `broadcast`), calls `torch.xpu.synchronize()` to make the copy visible before the collective, then issues one `broadcast` per rank per output tensor with `src=self.ranks[root]`. Supports both the single-tensor and list-of-tensors input forms; requires `sizes` and a pre-allocated `output` (same constraint we already impose in DP-attention call sites).

No changes to the NCCL/CUDA code path, no changes outside this function pair, no new imports (`_is_xpu` is already defined at module scope).

## Accuracy Tests

Verified on 2× Intel XPU with DP-attention enabled:
- End-to-end greedy decode matches CPU reference on a small prompt set (Qwen-family model, TP=1, DP=2).
- Before this change: crash inside `reduce_scatterv` / `all_gatherv` because `pynccl_comm` is `None`/disabled.
- After this change: run completes; per-token logits match single-device reference within bf16 tolerance.

## Speed Tests and Profiling

Not applicable to non-XPU backends — CUDA/ROCm/NPU paths are unchanged (verified by diffing the function bodies against `main`). On XPU, this is an enablement change (previously crashed), so there's no prior baseline to compare against; further tuning of the coalesced broadcast pattern can be a follow-up.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

cc: @siju-samuel @rbabukv

---
### CI States

Latest PR Test (Base): :x: [Run #30248463131](https://github.com/sgl-project/sglang/actions/runs/30248463131)
Latest PR Test (Extra): :x: [Run #30248463052](https://github.com/sgl-project/sglang/actions/runs/30248463052)

### CODE REVIEW

File: python/sglang/srt/distributed/parallel_state.py [hunk 1]
```diff
@@ -1128,6 +1152,53 @@ def all_gatherv(
...
+                    for root, split_size in enumerate(sizes):
+                        dst_slice = out[
+                            split_offset : split_offset + split_size
+                        ]
+                        torch.distributed.broadcast(
+                            dst_slice,
+                            src=self.ranks[root],
+                            group=self.device_group,
+                        )
+                        split_offset += split_size
```
Comment (siju-samuel) [inline python/sglang/srt/distributed/parallel_state.py]: replace this with this.  the coleasing manager might hang due to some bugs in xccl/oneccl
```
            for out in outputs:
                split_offset = 0
                for root, split_size in enumerate(sizes):
                    dst_slice = out[split_offset : split_offset + split_size]
                    torch.distributed.broadcast(
                        dst_slice,
                        src=self.ranks[root],
                        group=self.device_group,
                    )
                    split_offset += split_size

            return outputs
```

File: python/sglang/srt/distributed/parallel_state.py [hunk 2]
```diff
@@ -1128,6 +1152,53 @@ def all_gatherv(
...
+            )
+            assert len(sizes) == world_size
+            my_offset = sum(sizes[: self.rank_in_group])
+            my_size = sizes[self.rank_in_group]
+
+            for inp, out in zip(inputs, outputs):
+                assert inp.shape[0] == my_size
+                assert out is not None
+                out[my_offset : my_offset + my_size].copy_(inp)
+            torch.xpu.synchronize()
```
Comment (siju-samuel) [inline python/sglang/srt/distributed/parallel_state.py]: Remove the xpu.synchronize() we do not need full device sync

File: python/sglang/srt/distributed/parallel_state.py [hunk 3]
```diff
@@ -923,6 +923,30 @@ def reduce_scatterv(
...
+
+            if output is None:
+                output = torch.empty(
+                    output_shape, dtype=input_.dtype, device=input_.device
+                )
+            else:
+                assert output.shape == output_shape
+
+            torch.distributed.all_reduce(input_, group=self.device_group)
+            output.copy_(input_[my_offset : my_offset + chunk_size])
```
Comment (siju-samuel) [inline python/sglang/srt/distributed/parallel_state.py]: `reduce_scatterv `mutates input_ in place
so the caller's input buffer is silently overwritten with the globally-summed tensor. Any call site that reuses its input after reduce_scatterv would get corrupted data

File: python/sglang/srt/distributed/parallel_state.py [hunk 4]
```diff
@@ -1128,6 +1152,53 @@ def all_gatherv(
...
+                            split_offset : split_offset + split_size
+                        ]
+                        torch.distributed.broadcast(
+                            dst_slice,
+                            src=self.ranks[root],
+                            group=self.device_group,
+                        )
+                        split_offset += split_size
+
+            return outputs
```
Comment (siju-samuel) [inline python/sglang/srt/distributed/parallel_state.py]: Add some tests as well for both allgatherv and reducescaterv

**Review by siju-samuel — CHANGES_REQUESTED** [review]
